src/librustdoc/clean/types.rs

```diff
@@ -1393,33 +1417,46 @@ impl Type {
     }
 }
 
-impl GetDefId for Type {
-    fn def_id(&self) -> Option<DefId> {
-        match *self {
-            ResolvedPath { did, .. } => Some(did),
-            Primitive(p) => cache().primitive_locations.get(&p).cloned(),
-            BorrowedRef { type_: box Generic(..), .. } => {
-                Primitive(PrimitiveType::Reference).def_id()
-            }
-            BorrowedRef { ref type_, .. } => type_.def_id(),
+impl Type {
+    fn inner_def_id(&self, cache: Option<&Cache>) -> Option<DefId> {
+        let t: &dyn GetDefId = match *self {
+            ResolvedPath { did, .. } => return Some(did),
+            Primitive(p) => return cache.and_then(|c| c.primitive_locations.get(&p).cloned()),
+            BorrowedRef { type_: box Generic(..), .. } => &Primitive(PrimitiveType::Reference),
+            BorrowedRef { ref type_, .. } => return type_.inner_def_id(cache),
             Tuple(ref tys) => {
                 if tys.is_empty() {
-                    Primitive(PrimitiveType::Unit).def_id()
+                    &Primitive(PrimitiveType::Unit)
                 } else {
-                    Primitive(PrimitiveType::Tuple).def_id()
+                    &Primitive(PrimitiveType::Tuple)
                 }
             }
-            BareFunction(..) => Primitive(PrimitiveType::Fn).def_id(),
-            Never => Primitive(PrimitiveType::Never).def_id(),
-            Slice(..) => Primitive(PrimitiveType::Slice).def_id(),
-            Array(..) => Primitive(PrimitiveType::Array).def_id(),
-            RawPointer(..) => Primitive(PrimitiveType::RawPointer).def_id(),
-            QPath { ref self_type, .. } => self_type.def_id(),
-            _ => None,
+            BareFunction(..) => &Primitive(PrimitiveType::Fn),
+            Never => &Primitive(PrimitiveType::Never),
+            Slice(..) => &Primitive(PrimitiveType::Slice),
+            Array(..) => &Primitive(PrimitiveType::Array),
+            RawPointer(..) => &Primitive(PrimitiveType::RawPointer),
+            QPath { ref self_type, .. } => return self_type.inner_def_id(cache),
+            // FIXME: remove this wildcard
+            _ => return None,
+        };
```

Comment:
This now uses a Primitive everywhere I think, so you can get rid of `&dyn GetDefId` and just use a single `Primitive` type.